### PR TITLE
Migrate linting to ESLint

### LINKED ISSUE #529: Migrate linting tools to ESLint
We're currently using both [JSHint](https://github.com/jshint/jshint) and [JSCS](https://github.com/jscs-dev/node-jscs) to lint our Javascript files to check linting and enforce code style (which we're not doing too much of as of yet, see #209).  Migrating to ESLint would both reduce the number of tools we use, as well as reduce linting time as JSHint and JSCS need to both independently read and parse the Javascript files, thereby doubling the parse time.

### PR DESCRIPTION
**Pull request by nightpool**
JSHint is less active, less well-maintained, less supported, and less customizable.

Plus, eslint lets us handle style lints as well, as mentioned by issue #529

I'm super wiped, but justification of style choices TK.

fixes #529, probably also #1099.

version bumps should probably be added before merge

The important commits to look at are a1dd709, 10f5c3c, and fc3fa90. The others are mostly automated. Try reviewing with whitespace disabled (`?w=1` at the end of a github url)

Main non-format lints that are new:

 - `no-undef`: all variables must be var-defined in a scope, they can't be implicitly global. Global variables must be declared either in `.eslintrc` (for ones used often) or in an inline comment (with an explanation of where they're declared)
 - `no-unused`: no unused variables
 - `id-length`: identifiers must be at least 2 characters. I would prefer three, but there are a few 2-letter words that are useful and I didn't want to go through making exceptions for all of them
 - `no-shadow`: identifiers must not shadow variables in other scopes. The resolution applied here was sometimes to change the name, and, in a few places, use let-bindings to narrow the scope. (var  bindings are hoisted to the enclosing function, let bindings have block-based scope)
 - `valid-jsdocs` any function that has jsdocs now has those validated for having param and return statements for any params or return values as well as jsdoc syntax errors

suggestions for other eslint rules that should be applied are welcome, but they may need to wait for a followup pr

new format lints applied:

 - tab indentation
 - always indent blocks (there were places in XKit where we weren't doing this, and it was very confusing)
 - spaces after keywords, spaces after function definitions.
    ```js
    if (cond) {
        // statement
    }

    function bar() {
        // statement
    }
    ```
 - unix linebreaks
 - space between infix operators (`1 + 2`)
 - space after commas

### CODE REVIEW

File: .eslintrc.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      - before: true
+        after: true
+  globals:
+    require: true
+    unsafeWindow: true
+
+    XKit: true
+    XBridge: true
+    Tumblr: true
+    add_tag: true
```
Comment (nightpool) [inline .eslintrc.yml]: Tumblr and add_tag being here are a compromise between explicitness and usability. I would prefer that we would have to inline declare them as globals in every function that gets injected into the browser scope, but a) this is somewhat impractical and b) the way eslint directives work the protection provided by doing it the other way is minimal

File: Extensions/activity_plus.js [hunk 2]
```diff
@@ -52,8 +55,6 @@ XKit.extensions.activity_plus = new Object({
 		}
 	},
 
-	new_note_check_interval: 0,
```
Comment (nightpool) [inline Extensions/activity_plus.js]: this was unused, and wasn't even being set correctly to begin with

File: Extensions/one_click_reply.js [hunk 3]
```diff
@@ -646,15 +647,6 @@ XKit.extensions.one_click_reply = new Object({
 
 		obj = $(obj).parent();
 
-		if ($(obj).hasClass("note") === false || $(obj).hasClass("action") === true) {
-			// Must be in a sub-div.
-			if ($(obj).hasClass("action") === true) {
-				n_box = $(obj).parent();
```
Comment (nightpool) [inline Extensions/one_click_reply.js]: n_box is unused and undefined here

File: Extensions/audio_downloader.js [hunk 4]
```diff
@@ -129,7 +129,7 @@ XKit.extensions.audio_downloader = new Object({
 		var text = "";
 		var possible = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
 
-		for( var i=0; i < 15; i++ )
+		for ( var i = 0; i < 15; i++ )
```
Comment (nightpool) [inline Extensions/audio_downloader.js]: TODO: fix this

**Comment by nightpool** [comment]
I think this is ready now, pending a few small cleanup things. Can someone please review? Thanks!

File: Firefox/index.js [hunk 5]
```diff
@@ -1,115 +1,115 @@
...
+	});
 
-  port.on('reset_all', function(data) {
-    console.log('reset_all()');
-    let existingKeys = prefs.keys(prefRoot);
-    for (let name of existingKeys) {
-      prefs.reset(name);
-    }
-  });
+	port.on('reset_all', function(_data) {
```
Comment (hobinjk) [inline Firefox/index.js]: Does this need to be _data?

File: xkit.js [hunk 6]
```diff
@@ -128,889 +167,828 @@ XKit = {
 			if (!xkit_main.errors && xkit_main.script) {
 				XKit.console.add("Trying to run xkit_main.");
 				try {
-					/* jshint evil: true */
 					eval(xkit_main.script + "\n//# sourceURL=xkit/xkit_main.js");
+					XKit.frame_mode = true;
```
Comment (hobinjk) [inline xkit.js]: Where does this come from?

**Comment by hobinjk** [comment]
This is a heroic effort! It looks like so many bugs have been fixed.
✨ ✨ ✨ ✨

**Comment by nightpool** [comment]
@hobinjk XKit 7.7.3 doesn't actually work on safari 9 anyway, so I figured
we were probably good there. Also, it's very small percentage wise. I don't
know about our specific demographics, but Apple has been pretty aggressive
about updating it even on non-Sierra systems.

File: Firefox/index.js (hunk 5 above)
Comment (nightpool) [inline Firefox/index.js]: No, whoops, I meant to undo that change. I was tinkering with having unused function parameters be required to be named with a leading underscore, but I decided it wasn't that useful in the long run. Maybe in a followup PR.

File: Extensions/xkit_main.js [hunk 7]
```diff
@@ -2,276 +2,279 @@
...
 
-		// We'll be running patches first.
-		if (extension_id === "xkit_patches" && force !== true) { XKit.extensions.xkit_main.run_next_extension(); return; }
+			var xkit_main = XKit.installed.get(extension_id);
 
-		var xkit_main = XKit.installed.get(extension_id);
+			// Check if in Frame Mode.
+			if (XKit.frame_mode === true && extension_id !== "xkit_patches") {
+				// This is ugly: I don't want to eval script.
+				eval(xkit_main.script + "\n	//# sourceURL=xkit/" + extension_id + ".js");
```
Comment (nightpool) [inline Extensions/xkit_main.js]: gc.

File: xkit.js (hunk 6 above)
Comment (nightpool) [inline xkit.js]: It didn't change? 
image

**Comment by nightpool** [comment]
addressed the non-let stuff.

**Comment by hobinjk** [comment]
Okay, I deleted all the `let` comments. I don't think you addressed the comment on line 332 of `profiler.js`

File: Extensions/profiler.js [hunk 8]
```diff
@@ -319,17 +319,17 @@ XKit.extensions.profiler = new Object({
...
 		} else {
 			$("#xkit-profiler-nickname").removeClass("loading-up").html(m_nickname);
 		}
 
 		var m_blogs = XKit.tools.get_blogs();
-		for(i=0;i<m_blogs.length;i++) {
+		for (var i = 0; i < m_blogs.length; i++) {
 			if (m_blogs[i] !== "") {
-				blog_id = m_blogs[i];
+				var blog_id = m_blogs[i];
```
Comment (nightpool) [inline Extensions/profiler.js]: something something hoisting something something.

**Comment by nightpool** [comment]
(@hobinjk addressed, sorry)

**Comment by hobinjk** [comment]
Testing this further I'm having issues with the current shrinkwrapped dependency versions and Node 7.4.0. Recreating npm-shrinkwrap.json seems to fix the crash on 'internal/fs' even though there's still a Node-7-incompatible version of graceful-fs present in the tree.

**Comment by nightpool** [comment]
Oops, that's probably my fault. My node install is completely FUBAR'd. I'm pretty sure I'm on v6. Can you fix the shrinkwrap and push?

**Comment by hobinjk** [comment]
Pushed!

I tested locally and nothing seems broken. I do want to shove through as many PRs as possible before this gets merged because this is definitely going to cause some merge conflicts. I should have time Soon™️ to dedicate to the cause.

**Comment by nightpool** [comment]
@hobinjk done.

**Comment by hobinjk** [comment]
@homu r+
time to watch the :fire:

**Pull request merged**